src/d2lfetch-dedupe.js

```diff
@@ -1,51 +1,116 @@
+ */
+
 export class D2LFetchDedupe {
 
 	constructor() {
-		this._inflightRequests = this._inflightRequests || [];
+		/** @type {InflightRequestInfo} */
+		this._inflightRequests = {};
+		this._nextReqId = 0;
 	}
 
+	/**
+	 * @param {Request} request
+	 * @param {MiddlewareFunc} next
+	 *
+	 * @returns {Promise<Response>}
+	 */
 	dedupe(request, next) {
 		if (false === request instanceof Request) {
 			return Promise.reject(new TypeError('Invalid request argument supplied; must be a valid window.Request object.'));
 		}
 
 		if (request.method !== 'GET' && request.method !== 'HEAD' && request.method !== 'OPTIONS') {
-			if (!next) {
-				return Promise.resolve(request);
-			}
 			return next(request);
 		}
 
 		const key = this._getKey(request);
-		if (this._inflightRequests[key]) {
-			this._inflightRequests[key].count++;
-			return this._inflightRequests[key].action;
-		}
+		const reqId = this._nextReqId++;
+
+		let newRequest = false;
+
+		if (!this._inflightRequests[key]) {
+			const abortController = new AbortController();
+			const inflightRequest = new Request(request, {
+				signal: abortController.signal
+			});
 
-		if (!next) {
-			return Promise.resolve(request);
+			this._inflightRequests[key] = {
+				inflightRequest,
+				abortController,
+				dedupedRequests: {}
+			};
+
+			newRequest = true;
 		}
 
-		const result = next(request);
-		if (result && result instanceof Promise) {
-			this._inflightRequests[key] = { count: 1 };
-			this._inflightRequests[key].action = result
-				.then(function(response) {
-					const usedMultiple = this._inflightRequests[key].count !== 1;
-					delete this._inflightRequests[key];
-					return this._clone(response, usedMultiple);
-				}.bind(this))
-				.catch(function(err) {
-					delete this._inflightRequests[key];
-					return Promise.reject(err);
-				}.bind(this));
-
-			return this._inflightRequests[key].action;
+		const dedupedRequest = new Promise((resolve, reject) => {
+			this._inflightRequests[key].dedupedRequests[reqId] = { resolve, reject };
+
+			if (request.signal && typeof request.signal.addEventListener === 'function') {
+				request.signal.addEventListener('abort', () => {
+					const inflightRequest = this._inflightRequests[key];
+
+					if (Object.keys(inflightRequest.dedupedRequests).length === 1) {
+						const abortController = this._inflightRequests[key].abortController;
+
+						if (abortController && typeof abortController.abort === 'function') {
+							abortController.abort();
+						}
+					}
+
+					delete inflightRequest.dedupedRequests[reqId];
+					reject(new DOMException('Request was aborted.', 'AbortError'));
+				});
+			}
+		});
+
+		if (!newRequest) {
+			return dedupedRequest;
 		}
 
-		return result;
+		next(this._inflightRequests[key].inflightRequest)
+			.then((response) => {
+				const dedupedRequests = Object.values(this._inflightRequests[key].dedupedRequests);
+				const res = dedupedRequests.length > 1
+					? this._clone(response)
+					: response;
```

Comment:
Maintaining previous behaviour, but it doesn't seem desirable to change the capabilities of the response object based on whether multiple requests were mapped to a single fetch.